src/aggregator/client/client.go

```diff
@@ -364,4 +468,169 @@ func (c *client) writeTimeRangeFor(shard shard.Shard) (int64, int64) {
+	payload payloadUnion,
+) error {
+	m.shard = shard
+
+	switch payload.payloadType {
+	case untimedType:
+		switch payload.untimed.metric.Type {
+		case metric.CounterType:
+			value := unaggregated.CounterWithMetadatas{
+				Counter:         payload.untimed.metric.Counter(),
+				StagedMetadatas: payload.untimed.metadatas,
+			}
+			if err := value.ToProto(&m.cm); err != nil {
+				return err
+			}
+
+			m.metric = metricpb.MetricWithMetadatas{
+				Type:                 metricpb.MetricWithMetadatas_COUNTER_WITH_METADATAS,
+				CounterWithMetadatas: &m.cm,
+			}
+		case metric.TimerType:
+			value := unaggregated.BatchTimerWithMetadatas{
+				BatchTimer:      payload.untimed.metric.BatchTimer(),
+				StagedMetadatas: payload.untimed.metadatas,
+			}
+			if err := value.ToProto(&m.bm); err != nil {
+				return err
+			}
+
+			m.metric = metricpb.MetricWithMetadatas{
+				Type:                    metricpb.MetricWithMetadatas_BATCH_TIMER_WITH_METADATAS,
+				BatchTimerWithMetadatas: &m.bm,
+			}
+		case metric.GaugeType:
+			value := unaggregated.GaugeWithMetadatas{
+				Gauge:           payload.untimed.metric.Gauge(),
+				StagedMetadatas: payload.untimed.metadatas,
+			}
+			if err := value.ToProto(&m.gm); err != nil {
+				return err
+			}
+
+			m.metric = metricpb.MetricWithMetadatas{
+				Type:               metricpb.MetricWithMetadatas_GAUGE_WITH_METADATAS,
+				GaugeWithMetadatas: &m.gm,
+			}
+		default:
+			return fmt.Errorf("unrecognized metric type: %v",
+				payload.untimed.metric.Type)
+		}
+	case forwardedType:
+		value := aggregated.ForwardedMetricWithMetadata{
+			ForwardedMetric: payload.forwarded.metric,
+			ForwardMetadata: payload.forwarded.metadata,
+		}
+		if err := value.ToProto(&m.fm); err != nil {
+			return err
+		}
+
+		m.metric = metricpb.MetricWithMetadatas{
+			Type:                        metricpb.MetricWithMetadatas_FORWARDED_METRIC_WITH_METADATA,
+			ForwardedMetricWithMetadata: &m.fm,
+		}
+	case timedType:
+		value := aggregated.TimedMetricWithMetadata{
+			Metric:        payload.timed.metric,
+			TimedMetadata: payload.timed.metadata,
+		}
+		if err := value.ToProto(&m.tm); err != nil {
+			return err
+		}
+
+		m.metric = metricpb.MetricWithMetadatas{
+			Type:                    metricpb.MetricWithMetadatas_TIMED_METRIC_WITH_METADATA,
+			TimedMetricWithMetadata: &m.tm,
+		}
+	default:
+		return fmt.Errorf("unrecognized payload type: %v",
+			payload.payloadType)
+	}
+
+	size := m.metric.Size()
+	if size > cap(m.buf) {
+		const growthFactor = 1.5
+		m.buf = make([]byte, int(growthFactor*float64(size)))
+	}
+
+	// Resize buffer to exactly how long we need for marshalling.
+	m.buf = m.buf[:size]
+
+	_, err := m.metric.MarshalTo(m.buf)
+	return err
+}
+
+func (m *message) Shard() uint32 {
+	return m.shard
+}
+
+func (m *message) Bytes() []byte {
+	return m.buf
+}
+
+func (m *message) Size() int {
+	return len(m.buf)
+}
+
+func (m *message) Finalize(reason producer.FinalizeReason) {
+	// Return to pool.
```

Comment:
Do we need to reset any of proto structs `m.metric`, `m.tm`, `m.gm` etc?

Reply:
Thankfully they all get reset just before use in Encode so no need to do it twice.